Make test_export training IR compatible

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138517

In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

tugsbayasgalan committed:
Make test_export training IR compatible

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -870,7 +876,7 @@ def forward(self, x):
                 return x + y + z + w2
 
-        ep = export(M(), (torch.randn(2, 3),), strict=False)
+        ep = export(M(), (torch.randn(2, 3),), strict=False).run_decompositions({})
```
Why these extra `run_decompositions`? I guess I'm missing context on this PR, what does it mean to make these tests "training IR compatible"? Please add a summary to the PR.

tugsbayasgalan replied:
The test expects to see mutated buffer as extra output. Only way we get it today is via functiionalizaiton.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -1160,47 +1166,6 @@ def forward(self, x):
         )
         check_users_for_graph(ep.graph)
 
-    def test_export_predispatch_custom_ops_warnings(self):
```
It doesn't matter anymore

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -3955,7 +3920,7 @@ class Module(torch.nn.Module):
                 return x.to("cpu")
 
-        ep = export(Module(), (torch.tensor(1, device="cpu"),))
+        ep = export(Module(), (torch.tensor(1, device="cpu"),)).run_decompositions({})
```
Tests expect to see functional operator.

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -5545,7 +5461,7 @@ def forward(self, x):
         self.assertTrue(torch.allclose(unflattened(*inps), M2()(*inps)))
 
-    @testing.expectedFailureRetraceability  # Retracing tensor constants results in buffers
+    #@testing.expectedFailureRetraceability  # Retracing tensor constants results in buffers
```
So this goes away, is that good or bad? Just want to make sure that if there's still something to fix, we keep an expected failure.

tugsbayasgalan replied:
The failure was actually not related to tensor constants lol. It was just some graph mismatch.

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -4257,71 +4236,18 @@ def forward(self, x):
-            self.assertExpectedInline(
-                str(ep.graph_module.code).strip(),
-                """\
+        ep = export_for_training(M(), (torch.randn(2, 2),)).run_decompositions(
```
looks like you need to remove the xfail from here

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Make test_export training IR compatible"


In this PR, I make test_export to be compatible with training IR. The idea is that when we flip the IR to non-functional training IR, all these tests should be green. The changes involve reading through the test case, and add necessary decomposition etc to make sure the tests pass. For example, if the tests expect to see mutated buffers returned, we need to get them via running run_decomp. 

Differential Revision: [D64732360](https://our.internmc.facebook.com/intern/diff/D64732360)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64732360).
